wildmatch.go

```diff
@@ -183,7 +192,7 @@ func (w *Wildmatch) parseTokens(dirs []string) []token {
 			return tokens
 		}
 	}
-	components := w.parseTokensSimple(dirs[0:])
+	components := w.parseTokensSimple(dirs[0:endItem])
```

Comment:
I noticed first we could probably have dropped the `0:` in the previous round of edits, and then I began to wonder if `dirs` could just be mutated instead of managing the `endItem` index.  Here's what I came up with; see if this makes any sense to you:
```go
	var finalComponents []token

	if !w.gitattributes {
		trailingIsEmpty := len(dirs) > 1 && dirs[len(dirs)-1] == ""
		numNonEmptyDirs := len(dirs)
		if trailingIsEmpty {
			numNonEmptyDirs -= 1
		}
		if w.contents {
			finalComponents = []token{&trailingComponents{}}
			if trailingIsEmpty {
				// Strip off the trailing empty string.
				dirs = dirs[:numNonEmptyDirs]
			}
		}
		// If we have one component, ignoring trailing empty
		// components and we know that a directory is permissible....
		if numNonEmptyDirs == 1 && (trailingIsEmpty || w.contents) {
			// We don't have a slash in the middle, so this can go
			// anywhere in the hierarchy.  If there had been a slash
			// here, it would have been anchored at the root.
			rest := w.parseTokensSimple(dirs)
			tokens := append([]token{&unanchoredDirectory{
				Until: rest[0],
			}})
			// If we're not matching all contents, then do include
			// the empty component so we don't match
			// non-directories.
			if finalComponents == nil && len(rest) > 1 {
				finalComponents = rest[1:]
			}
			return append(tokens, finalComponents...)
		}
	}
	components := w.parseTokensSimple(dirs)
	return append(components, finalComponents...)
```

Reply:
I've gone ahead and adopted this with a co-author credit.